doc/en/user/source/extensions/mapml/template.rst

```diff
@@ -0,0 +1,199 @@
+
+An example mapml-feature.ftl file to modify a point layer would look like::
+    
+    <mapml- xmlns="http://www.w3.org/1999/xhtml">
+      <map-head>
+      </map-head>
+      <map-body>
+        <map-feature>
+          <#list attributes as attribute>
+            <#if attribute.name == "NAME">
+              <map-properties name="UPDATED ${attribute.name}" value="CHANGED ${attribute.value}"/>
+            <#else>
+              <map-properties name="${attribute.name}" value="${attribute.value}"/>
+            </#if>
+          </#list>
+          <#list attributes as gattribute>
+            <#if gattribute.isGeometry>
+              <map-geometry>
+                <map-point>
+                  <map-coordinates><#list gattribute.rawValue.coordinates as coord>
+                    <#if coord?index == 0><map-span class="desired">${coord.x} ${coord.y}</map-span><#else>${coord.x} ${coord.y}</#if></#list></map-coordinates></map-point>
+                </map-geometry>
+             </#if>
+           </#list>
+          </map-feature>
+        </map-body>
+        </mapml->
+
+This would result in a MapML feature output body that would resemble:
+    
+    <mapml- xmlns="http://www.w3.org/1999/xhtml"><map-head><map-title>Bridges</map-title><map-meta charset="UTF-8"/><map-meta content="text/mapml" http-equiv="Content-Type"/><map-meta name="extent" content="top-left-longitude=-180.000000,top-left-latitude=90.000000,bottom-right-longitude=180.000000,bottom-right-latitude=-90.000000"/><map-meta name="cs" content="gcrs"/><map-meta name="projection" content="MapML:urn:ogc:def:crs:EPSG::4326"/><map-style>.bbox {display:none} .Bridges-r1-s1{r:48.0; stroke-opacity:1.0; stroke-dashoffset:0; well-known-name:square; stroke-width:1.0; opacity:1.0; fill:#808080; fill-opacity:1.0; stroke:#000000; stroke-linecap:butt}</map-style></map-head><map-body><map-feature id="Bridges.1107531599613" class="Bridges-r1-s1"><map-geometry><map-point><map-coordinates><map-span class="desired">0.0007 0.0002</map-span></map-coordinates></map-point></map-geometry><map-properties><table xmlns="http://www.w3.org/1999/xhtml"><thead><tr><th role="columnheader" scope="col">Property name</th><th role="columnheader" scope="col">Property value</th></tr></thead><tbody><tr><th scope="row">FID</th><td itemprop="FID">110</td></tr><tr><th scope="row">UPDATED NAME</th><td itemprop="NAME">CHANGED Cam Bridge</td></tr></tbody></table></map-properties></map-feature></map-body></mapml->
+
+Note that in addition to tagging the coordinates with a style class, the template also changes the name of the NAME property to "UPDATED NAME" and the value to "CHANGED Cam Bridge".  
+
+A more complex example mapml-feature.ftl that operates on a multi-polygon layer and only inserts map-span tags into a specific feature::
+
+    <mapml- xmlns=\"http://www.w3.org/1999/xhtml\">
```

Comment:
I tried the previous examples and got them to work, +/- any comments above.  I think this example is the most complex so it would be ideal to make it work for one of the release layers, such as poly_landmarks.  So far I haven't got it to have a visible effect though.  One barrier is having to restart the server after each change to an .ftl file.

Reply:
> I tried the previous examples and got them to work, +/- any comments above. I think this example is the most complex so it would be ideal to make it work for one of the release layers, such as poly_landmarks. So far I haven't got it to have a visible effect though. One barrier is having to restart the server after each change to an .ftl file.

I just pushed some udpates that should help:
1)There was an issue with the non-geospatial attributes not getting updated that should be resolved.
2)I have added several more examples, covering all of the geometry types
3)I fixed an issue with geometrycollection parsing and added an example to the documentation.

With regards to the requirement to restart when the template is updated - that is required for performance reasons.  The template gets cached after the first time it gets used.  I can change that but there will be a performance penalty.  A possible compromise would be to hash the template and then monitor for changes - another performance penalty, but smaller than replacing it with every call.

Reply:
> With regards to the requirement to restart when the template is updated - that is required for performance reasons.

I can live with it.  It would be great if there was a debugging window in GeoServer where you could paste a ftl and see if it worked before saving it somewhere. 

I tried this template from your updated examples with poi:

```ftl
<mapml- xmlns="http://www.w3.org/1999/xhtml">
  <map-head>
  </map-head>
  <map-body>
    <map-feature>
      <#list attributes as attribute>
        <#if attribute.isGeometry>
          <map-geometry>
            <map-geometrycollection>
              <map-linestring><map-coordinates><#list attribute.rawValue.coordinates as coord>${coord.x} ${coord.y}</#list></map-coordinates></map-linestring>
              <map-point><map-coordinates><#list attribute.rawValue.coordinates as coord>${coord.x} ${coord.y}</#list></map-coordinates></map-point>
            </map-geometrycollection>
          </map-geometry>
        </#if>
      </#list>
    </map-feature>
  </map-body>
</mapml->
```

And I get this exception:

```xml
<?xml version="1.0" encoding="UTF-8"?><ServiceExceptionReport version="1.3.0" xmlns="http://www.opengis.net/ogc" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/ogc http://localhost:8080/geoserver/schemas/wms/1.3.0/exceptions_1_3_0.xsd">   <ServiceException>
      javax.xml.bind.UnmarshalException
 - with linked exception:
[Exception [EclipseLink-25004] (Eclipse Persistence Services - 2.7.4.v20190115-ad5b7c6b2a): org.eclipse.persistence.exceptions.XMLMarshalException
Exception Description: An error occurred unmarshalling the document
Internal Exception: org.xml.sax.SAXParseException; lineNumber: 9; columnNumber: 15; unexpected element (uri:&quot;http://www.w3.org/1999/xhtml&quot;, local:&quot;map-point&quot;). Expected elements are &lt;{http://www.w3.org/1999/xhtml}point&gt;,&lt;{http://www.w3.org/1999/xhtml}map-linestring&gt;,&lt;{http://www.w3.org/1999/xhtml}map-polygon&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multipoint&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multilinestring&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multipolygon&gt;]
Exception Description: An error occurred unmarshalling the document
Internal Exception: org.xml.sax.SAXParseException; lineNumber: 9; columnNumber: 15; unexpected element (uri:&quot;http://www.w3.org/1999/xhtml&quot;, local:&quot;map-point&quot;). Expected elements are &lt;{http://www.w3.org/1999/xhtml}point&gt;,&lt;{http://www.w3.org/1999/xhtml}map-linestring&gt;,&lt;{http://www.w3.org/1999/xhtml}map-polygon&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multipoint&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multilinestring&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multipolygon&gt;
unexpected element (uri:&quot;http://www.w3.org/1999/xhtml&quot;, local:&quot;map-point&quot;). Expected elements are &lt;{http://www.w3.org/1999/xhtml}point&gt;,&lt;{http://www.w3.org/1999/xhtml}map-linestring&gt;,&lt;{http://www.w3.org/1999/xhtml}map-polygon&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multipoint&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multilinestring&gt;,&lt;{http://www.w3.org/1999/xhtml}map-multipolygon&gt;
</ServiceException></ServiceExceptionReport>
```

I like the fact that the examples are tied to specific release layers so I can work from an example that I know is supposed to work (my freemarker skills are what any beginner would be capable of).